### PR TITLE
Kconfig: Additional STM32L4 Chips

### PR DESCRIPTION
**Pull request by Kmart777**
Added new STM32L4 devices.

### CODE REVIEW

File: nuttx/Kconfig [hunk 1]
```diff
@@ -311,10 +311,10 @@ config CHIP_MFG_ID
 		Chip Vendor ID.  For example STM's ID is 00000104
 
 config CHIP_PRODUCT_ID
-	hex "Unique Product ID for chip"
-	default 0x00006435
+	hex "Unique Product ID for chip (STM32L4X3 = 0x435, STM32L4X6 = 0x415)"
+        default 0x415
```
Comment (jwylder1) [inline nuttx/Kconfig]: 0x0006415 if you want to change the default.

**Review by CaveMike — CHANGES_REQUESTED** [review]
I think changing CHIP_PRODUCT_ID and WRAP_CHIP_PRODUCT_ID will cause issues with flashing since the chip IDs will no longer match what is read from the STM's ID register.

**Comment by Kmart777** [comment]
Hello CaveMike, 
    Perhaps there is insider information that I am not aware of but the STM32L4x6 reference manual (RM0351) section 48.6.1 and the STM32L43x reference manual (RM0394) section 44.6.1 specifies the 12 LSB as being relevant to this CHIP_PRODUCT_ID where the next upper 4 bits are reserved so undefined.   The MDK uses a STM32L476MG so by definition its default should be 0x415, not 0x435 per these specs.  Is there another reason why the default should be 0x00006435?   I have to admit that I'm not clear on the differences between CHIP_PRODUCT_ID and WRAP_CHIP_PRODUCT_ID, but suspect they should be the same, unless there might be an APBE attribute.   
Thanks

**Comment by CaveMike** [comment]
Do we have different versions? Here is the doc that I have:
> RM0351
> Reference manual
> STM32L4x6 advanced ARM-based 32-bit MCUs
> 25 June 2014
> DocID 024597 Rev 0.1
> 1/1709
> 

Here is the section I have:
> 44.6.1 MCU device ID code
> The STM32L4x6 MCUs integrate an MCU ID code. This ID identifies the ST MCU part- number and the die revision. It is part of the DBG_MCU component and is mapped on the external PPB bus (see Section 44.16 on page 1683). This code is accessible using the JTAG debug port (4 to 5 pins) or the SW debug port (two pins) or by the user software. It is even accessible while the MCU is under system reset.
> 
> DBGMCU_IDCODE
> Address: 0xE004 2000
> Only 32-bits access supported. Read-only
> This code is read as 0x10006415
> 
> Bits31:16 REV_ID(15:0): Revisionidentifier
> This field indicates the revision of the device. 
> For example, it is read as 0x1000 for Revision 1.0.
> 
> Bits15:0 DEV_ID(15:0): Deviceidentifier 
> This field indicates the device and its revision.
> The device ID is 0x6415.

up_getchipid() returns DBGMCU_IDCODE[15:0] or 0x6415.

Correct, CHIP_PRODUCT_ID and WRAP_CHIP_PRODUCT_ID are the same for this discussion.

**Comment by Kmart777** [comment]
For STML4v476, I'm referencing version-5 of  RM0351 (DocID024597) page 1782/1830

48.6.1 MCU device ID code
The STM32L4x5/STM32L4x6 MCUs integrate an MCU ID code. This ID identifies the ST
MCU part-number and the die revision. It is part of the DBG_MCU component and is
mapped on the external PPB bus (see Section 48.16 on page 1794). This code is
accessible using the JTAG debug port (4 to 5 pins) or the SW debug port (two pins) or by
the user software. It is even accessible while the MCU is under system reset.
Only the DEV_ID(11:0) should be used for identification by the debugger/programmer tools.
DBGMCU_IDCODE
Address: 0xE004 2000
Only 32-bits access supported. Read-only

Bits 31:16 REV_ID[15:0] Revision identifier
This field indicates the revision of the device.
0x1000: Rev 1 for STM32L475xx/476xx/486xx devices
Rev A for STM32L496xx/4A6xx devices
0x1001: Rev 2 for STM32L475xx/476xx/486xx devices
Rev B for STM32L496xx/4A6xx devices
0x1003: Rev 3 for STM32L475xx/476xx/486xx devices
0x1007: Rev 4 for STM32L475xx/476xx/486xx devices
Bits 15:12 Reserved, must be kept at reset value. <------------------------
Bits 11:0 DEV_ID[11:0]: Device identifier
The device ID is:
0x461 for STM32L496xx/4A6xx devices
0x415 for STM32L475xx/476xx/486xx devices.

Bits 15:12 are specified reserved in the version-5 ref manual.

For STM32L433/31, I'm referencing version-3 of  RM0394 (DocID027295) page 1437/1472
44.6.1 MCU device ID code
The STM32L43xxx/44xxx/45xxx/46xxx MCUs integrate an MCU ID code. This ID identifies
the ST MCU part-number and the die revision. It is part of the DBG_MCU component and is
mapped on the external PPB bus (see Section 44.16 on page 1449). This code is
accessible using the JTAG debug port (4 to 5 pins) or the SW debug port (two pins) or by
the user software. It is even accessible while the MCU is under system reset.
Only the DEV_ID(11:0) should be used for identification by the debugger/programmer tools.
DBGMCU_IDCODE
Address: 0xE004 2000
Only 32-bits access supported. Read-only
Bits 31:16 REV_ID[15:0] Revision identifier
This field indicates the revision of the device.
0x1000: Rev A
0x1001: Rev Z
Others: Reserved
Bits 15:12 Reserved, must be kept at reset value.
Bits 11:0 DEV_ID[11:0]: Device identifier
The device ID is:
– 0x435 for STM32L43xxx and STM32L44xxx devices
– 0x462 for STM32L45xxx and STM32L46xxx devices

Again, Bits 15:12 are specified reserved in the version-3 ref manual.

**Comment by CaveMike** [comment]
Thanks, that explains the discrepancy although I am unclear on the bit order since the least-significant nibble is the one to ignore (e.g. '5').

I think we'll still have to continue using DBGMCU_IDCODE[15:0] though since the code is already deployed that way.

File: nuttx/arch/arm/Kconfig [hunk 2]
```diff
@@ -299,6 +299,17 @@ config ARCH_CHIP
 	default "str71x"	if ARCH_CHIP_STR71X
 	default "tsb"		if ARCH_CHIP_TSB
 
+
```
Comment (CaveMike) [inline nuttx/arch/arm/Kconfig]: Please don't move these around. They're in alphabetical order.

File: nuttx/arch/arm/src/stm32/Kconfig [hunk 3]
```diff
@@ -1235,156 +1926,166 @@ if STM32_ADC1
 menu "ADC1 Device Initialization"
 
 config STM32_ADC1_CHAN0_INIT
-	bool "ADC1 Channel 0"
+	bool "ADC1 Channel 0 (IN1-PC0)"
```
Comment (CaveMike) [inline nuttx/arch/arm/src/stm32/Kconfig:1929]: I don't think PC0 is the only pin for ADC1, channel 0. Is it?

Comment (Kmart777) [reply nuttx/arch/arm/src/stm32/Kconfig:1929]: Hi CaveMike,
     This section relates to ADC1, subsequent sections relate to ADC2 and ADC3.   PC0 is configurable for all ADC channels as IN1.
Port	ADC
PA0	ADC12_IN5
PA1	ADC12_IN6
PA2	ADC12_IN7
PA3	ADC12_IN8
PA4	ADC12_ IN9
PA5	ADC12_ IN10
PA6	ADC12_ IN11
PA7	ADC12_ IN12
PB0	ADC12_ IN15
PB1	ADC12_ IN16
PC0	ADC123_ IN1
PC1	ADC123_ IN2
PC2	ADC123_ IN3
PC3	ADC123_ IN4
PC4	ADC12_ IN13
PC5	ADC12_ IN14
PF3	ADC3_IN6
PF4	ADC3_IN7
PF5	ADC3_IN8
PF6	ADC3_IN9
PF7	ADC3_IN10
PF8	ADC3_IN11
PF9	ADC3_IN12
PF10	ADC3_IN13

Comment (CaveMike) [reply nuttx/arch/arm/src/stm32/Kconfig:1929]: Sorry, maybe I am not following how the Kconfig is arranged.

STM32_ADC1_CHAN0_INIT is shared by all flavors of STM32 chip, right? Not just STM32L4xx, but also STM32Fxxx, etc. In these cases, is PC0 always the pin?

STM32_ADC1_CHAN0_INIT is the config for the ADC1, channel 0 initialization function, but that function might physically be routed out on different pins depending on the flavor of the STM chip.

Comment (Kmart777) [reply nuttx/arch/arm/src/stm32/Kconfig:1929]: Hi CaveMike,
     I don't currently have the data to confirm chipsets outside the STM32L4 series, but the config scripts have not changed with the exception to adding the INx designation and the addition of "depends on" conditions within the ADC 1 to 3 lists that trim the list based on STM32L4 type and package size.   The following list the ADC delta between v476 and v433/31.  

Port	ADC for v476	ADC for v433/31
PA0	ADC12_IN5	ADC1_IN5
PA1	ADC12_IN6	ADC1_IN6
PA2	ADC12_IN7	ADC1_IN7
PA3	ADC12_IN8	ADC1_IN8
PA4	ADC12_ IN9	ADC1_IN9
PA5	ADC12_ IN10	ADC1_IN10
PA6	ADC12_ IN11	ADC1_IN11
PA7	ADC12_ IN12	ADC1_IN12
PB0	ADC12_ IN15	ADC1_IN15
PB1	ADC12_ IN16	ADC1_IN16
PC0	ADC123_ IN1	ADC1_IN1
PC1	ADC123_ IN2	ADC1_IN2
PC2	ADC123_ IN3	ADC1_IN3
PC3	ADC123_ IN4	ADC1_IN4
PC4	ADC12_ IN13	ADC1_IN13
PC5	ADC12_ IN14	ADC1_IN14
PF3	ADC3_IN6	
PF4	ADC3_IN7	
PF5	ADC3_IN8	
PF6	ADC3_IN9	
PF7	ADC3_IN10	
PF8	ADC3_IN11	
PF9	ADC3_IN12	
PF10	ADC3_IN13

File: nuttx/Kconfig (hunk 1 above)
Comment (flowflow25) [inline nuttx/Kconfig]: Remove this change until we can properly test it and ensure there are no conflicts with the AP.

File: nuttx/Kconfig [hunk 4]
```diff
@@ -329,16 +329,16 @@ config WRAP_CHIP_MFG_ID
 		the MuC.
 
 config WRAP_CHIP_PRODUCT_ID
-	hex "Unique Product ID for MuC Chip"
-	default 0x00006435
+	hex "Unique Product ID for MuC Chip  (STM32L4X3 = 0x6435, STM32L4X6 = 0x6415)"
+        default 0x6415
```
Comment (flowflow25) [inline nuttx/Kconfig]: Remove this changes also.

File: nuttx/arch/arm/Kconfig (hunk 2 above)
Comment (flowflow25) [inline nuttx/arch/arm/Kconfig]: Remove the added line.

**Review by flowflow25 — CHANGES_REQUESTED** [review]

**Review by jwylder1 — CHANGES_REQUESTED** [review]
Ken if you want to separate out the chipid change in nuttx/Kconfig, that is correct and can go in immediately.

**Comment by Kmart777** [comment]
Aborting this change after collaborating with JW and MC.

**Pull request closed without merging**